File: tests/test_swarm.py
import pytest
import asyncio
import json
import httpx
from unittest.mock import Mock, patch, MagicMock, AsyncMock
from zena_mode.arbitrage import SwarmArbitrator, ConsensusMethod

class TestSwarmArbitrator:
    @pytest.mark.asyncio
    @patch('httpx.AsyncClient')
    async def test_discover_swarm_enabled(self, mock_client_class):
        """Test discover_swarm detects live endpoints correctly."""
        # Mock client and its GET method
        mock_client = AsyncMock()
        mock_client_class.return_value.__aenter__.return_value = mock_client
        
        async def mock_get(url, timeout=None):
            mock_resp = Mock()
            if "8001/health" in url or "8006/health" in url:
                mock_resp.status_code = 200
            else:
                mock_resp.status_code = 404
            return mock_resp
            
        mock_client.get.side_effect = mock_get
        
        with patch('zena_mode.arbitrage.SWARM_ENABLED', True), \
             patch('zena_mode.arbitrage.SWARM_SIZE', 3):
            arb = SwarmArbitrator(ports=None) 
            await arb.discover_swarm()
            
        assert 8001 in arb.ports
        assert 8006 in arb.ports
        assert len(arb.ports) == 2
        assert len(arb.endpoints) == 2

    @pytest.mark.asyncio
    async def test_discover_swarm_disabled(self):
        """Test discover_swarm respects SWARM_ENABLED=False."""
        with patch('zena_mode.arbitrage.SWARM_ENABLED', False):
            arb = SwarmArbitrator(ports=None)
            await arb.discover_swarm()
            
        assert len(arb.ports) == 1
        assert arb.ports[0] == 8001

    def test_calculate_consensus_simple(self):
        """Test word-overlap consensus calculation."""
        arb = SwarmArbitrator(ports=[8001])
        
        # 100% Match
        assert arb._calculate_consensus_simple(["hello world", "hello world"]) == 1.0
        
        # 50% Match (hello is common, world/there are distinct)
        # common={hello}, union={hello, world, there} -> 1/3 = 0.333
        score = arb._calculate_consensus_simple(["hello world", "hello there"])
        assert 0.3 < score < 0.4
        
        # 0% Match
        assert arb._calculate_consensus_simple(["apple", "banana"]) == 0.0

    @pytest.mark.asyncio
    async def test_query_model_success(self):
        """Test _query_model handles successful HTTP responses."""
        arb = SwarmArbitrator(ports=[8001])
        mock_client = AsyncMock(spec=httpx.AsyncClient)
        mock_resp = Mock()
        mock_resp.status_code = 200
        mock_resp.json.return_value = {
            "choices": [{"message": {"content": "I am 95% confident this works."}}],
            "model": "Test-Model-v1"
        }
        mock_client.post.return_value = mock_resp
        
        result = await arb._query_model(mock_client, arb.endpoints[0], [{"role": "user", "content": "hi"}])
        
        assert "95% confident" in result["content"]
        assert result["model"] == "Test-Model-v1"
        assert result["confidence"] == 0.95

    @pytest.mark.asyncio
    async def test_query_model_error(self):
        """Test _query_model handles HTTP errors gracefully."""
        arb = SwarmArbitrator(ports=[8001])
        mock_client = AsyncMock(spec=httpx.AsyncClient)
        mock_resp = Mock()
        mock_resp.status_code = 500
        mock_client.post.return_value = mock_resp
        
        result = await arb._query_model(mock_client, arb.endpoints[0], [{"role": "user", "content": "hi"}])
        
        assert "Error: 500" in result["content"]

    @pytest.mark.asyncio
    async def test_get_cot_response_single_reflection(self):
        """Test CoT flow with a single model (Reflection mode)."""
        arb = SwarmArbitrator(ports=[8001])
        
        responses = [
            {"content": "Initial Thought", "time": 0.1, "model": "M1", "confidence": 0.8},
            {"content": "Refined Thought", "time": 0.1, "model": "M1", "confidence": 0.9}
        ]
        
        with patch.object(arb, '_query_model_with_timeout', side_effect=responses):
            class MockResp:
                async def aiter_lines(self):
                    yield 'data: {"choices": [{"delta": {"content": "Final Result"}}]}'
                    yield 'data: [DONE]'

            class MockContext:
                async def __aenter__(self): return MockResp()
                async def __aexit__(self, *args, **kwargs): pass

            class MockClient:
                async def __aenter__(self): return self
                async def __aexit__(self, *args, **kwargs): pass
                def stream(self, *args, **kwargs): return MockContext()
            
            with patch('httpx.AsyncClient', return_value=MockClient()):
                output = []
                async for chunk in arb.get_cot_response("test", "sys"):
                    output.append(chunk)
                
                full_text = "".join(output)
                assert "Final Result" in full_text

    def test_detect_contradictions(self):
        """Test embedding-based contradiction detection."""
        arb = SwarmArbitrator(ports=[8001])
        # Very different sentences should trigger contradiction
        responses = [
            "The capital of France is Paris.",
            "Sharks are a type of fruit that grows on trees."
        ]
        
        contradictions = arb.detect_contradictions(responses)
        assert len(contradictions) > 0
        assert contradictions[0]['pair'] == (1, 2)
        assert contradictions[0]['similarity'] < 0.2

    @pytest.mark.asyncio
    async def test_external_agent_bridge(self):
        """Test LiteLLM bridge placeholder."""
        arb = SwarmArbitrator(ports=[8001])
        result = await arb._query_external_agent("gpt-4o", [{"role": "user", "content": "hi"}])
        assert "LITELLM MOCK" in result["content"]
        assert result["model"] == "gpt-4o"

    def test_autogen_init(self):
        """Test AutoGen swarm initialization stub."""
        arb = SwarmArbitrator(ports=[8001])
        # Should not raise exception
        arb.init_autogen_swarm()

File: rust_output/src/test_swarm.rs
use anyhow::{Result, Context};
use crate::arbitrage::{SwarmArbitrator, ConsensusMetho};
use std::collections::HashMap;
use tokio;

/// TestSwarmArbitrator class.
#[derive(Debug, Clone)]
pub struct TestSwarmArbitrator {
}

impl TestSwarmArbitrator {
    /// Test discover_swarm detects live endpoints correctly.
    pub async fn test_discover_swarm_enabled(&self, mock_client_class: String) -> () {
        // Test discover_swarm detects live endpoints correctly.
        let mut mock_client = AsyncMock();
        mock_client_class.return_value.__aenter__.return_value = mock_client;
        let mock_get = |url, timeout| {
            // Mock get.
            let mut mock_resp = Mock();
            if (url.contains(&"8001/health".to_string()) || url.contains(&"8006/health".to_string())) {
                mock_resp.status_code = 200;
            } else {
                mock_resp.status_code = 404;
            }
            mock_resp
        };
        mock_client.get.side_effect = mock_get;
        /* let mock_config = mock::/* mock::patch(...) */ — use mockall crate */;
        {
            mock_config.swarm_enabled = true;
            mock_config.swarm_size = 3;
            mock_config.llm_port = 8001;
            mock_config.host = "127.0.0.1".to_string();
            let mut arb = SwarmArbitrator(/* ports= */ None);
            arb.discover_swarm().await;
        }
        assert!(arb.ports.contains(&8001));
        assert!(arb.ports.contains(&8006));
        assert!(arb.ports.len() == 2);
        assert!(arb.endpoints.len() == 2);
    }
    /// Test discover_swarm respects config::swarm_enabled=false.
    pub async fn test_discover_swarm_disabled(&self) -> () {
        // Test discover_swarm respects config::swarm_enabled=false.
        /* let mock_config = mock::/* mock::patch(...) */ — use mockall crate */;
        {
            mock_config.swarm_enabled = false;
            mock_config.llm_port = 8001;
            mock_config.host = "127.0.0.1".to_string();
            let mut arb = SwarmArbitrator(/* ports= */ None);
            arb.discover_swarm().await;
        }
        assert!(arb.ports.len() == 1);
        assert!(arb.ports[0] == 8001);
    }
    /// Test word-overlap consensus calculation.
    pub fn test_calculate_consensus_simple(&self) -> () {
        // Test word-overlap consensus calculation.
        let mut arb = SwarmArbitrator(/* ports= */ vec![8001]);
        assert!(arb._calculate_consensus_simple(vec!["hello world".to_string(), "hello world".to_string()]) == 1.0_f64);
        let mut score = arb._calculate_consensus_simple(vec!["hello world".to_string(), "hello there".to_string()]);
        assert!((0.3_f64 < score) && (score < 0.4_f64));
        assert!(arb._calculate_consensus_simple(vec!["apple".to_string(), "banana".to_string()]) == 0.0_f64);
    }
    /// Test _query_model handles successful HTTP responses.
    pub async fn test_query_model_success(&self) -> () {
        // Test _query_model handles successful HTTP responses.
        let mut arb = SwarmArbitrator(/* ports= */ vec![8001]);
        let mut mock_client = AsyncMock(/* spec= */ httpx.AsyncClient);
        let mut mock_resp = Mock();
        mock_resp.status_code = 200;
        mock_resp.json::return_value = HashMap::from([("choices".to_string(), vec![HashMap::from([("message".to_string(), HashMap::from([("content".to_string(), "I am 95% confident this works.".to_string())]))])]), ("model".to_string(), "Test-Model-v1".to_string())]);
        mock_client.post.return_value = mock_resp;
        let mut result = arb._query_model(mock_client, arb.endpoints[0], vec![HashMap::from([("role".to_string(), "user".to_string()), ("content".to_string(), "hi".to_string())])]).await;
        assert!(result["content".to_string()].contains(&"95% confident".to_string()));
        assert!(result["model".to_string()] == "Test-Model-v1".to_string());
        assert!(result["confidence".to_string()] == 0.95_f64);
    }
    /// Test _query_model handles HTTP errors gracefully.
    pub async fn test_query_model_error(&self) -> () {
        // Test _query_model handles HTTP errors gracefully.
        let mut arb = SwarmArbitrator(/* ports= */ vec![8001]);
        let mut mock_client = AsyncMock(/* spec= */ httpx.AsyncClient);
        let mut mock_resp = Mock();
        mock_resp.status_code = 500;
        mock_client.post.return_value = mock_resp;
        let mut result = arb._query_model(mock_client, arb.endpoints[0], vec![HashMap::from([("role".to_string(), "user".to_string()), ("content".to_string(), "hi".to_string())])]).await;
        assert!(result["content".to_string()].contains(&"Error: 500".to_string()));
    }
    /// Test CoT flow with a single model (Reflection mode).
    pub async fn test_get_cot_response_single_reflection(&self) -> () {
        // Test CoT flow with a single model (Reflection mode).
        let mut arb = SwarmArbitrator(/* ports= */ vec![8001]);
        let mut responses = vec![HashMap::from([("content".to_string(), "Initial Thought".to_string()), ("time".to_string(), 0.1_f64), ("model".to_string(), "M1".to_string()), ("confidence".to_string(), 0.8_f64)]), HashMap::from([("content".to_string(), "Refined Thought".to_string()), ("time".to_string(), 0.1_f64), ("model".to_string(), "M1".to_string()), ("confidence".to_string(), 0.9_f64)])];
        let _ctx = patch.object(arb, "_query_model_with_timeout".to_string(), /* side_effect= */ responses);
        {
            // TODO: nested class MockResp
            // TODO: nested class MockContext
            // TODO: nested class MockClient
            /* let _ctx = mock::/* mock::patch(...) */ — use mockall crate */;
            {
                let mut output = vec![];
                // async for
                while let Some(chunk) = arb.get_cot_response("test".to_string(), "sys".to_string()).next().await {
                    output.push(chunk);
                }
                let mut full_text = output.join(&"".to_string());
                assert!(full_text.contains(&"Final Result".to_string()));
            }
        }
    }
    /// Test embedding-based contradiction detection.
    pub fn test_detect_contradictions(&self) -> () {
        // Test embedding-based contradiction detection.
        let mut arb = SwarmArbitrator(/* ports= */ vec![8001]);
        let mut responses = vec!["The capital of France is Paris.".to_string(), "Sharks are a type of fruit that grows on trees.".to_string()];
        let mut contradictions = arb.detect_contradictions(responses);
        assert!(contradictions.len() > 0);
        assert!(contradictions[0]["pair".to_string()] == (1, 2));
        assert!(contradictions[0]["similarity".to_string()] < 0.2_f64);
    }
    /// Test external agent bridge when no API key is set.
    pub async fn test_external_agent_bridge_no_key(&self) -> () {
        // Test external agent bridge when no API key is set.
        let mut arb = SwarmArbitrator(/* ports= */ vec![8001]);
        let _ctx = patch.dict("os::environ".to_string(), HashMap::new(), /* clear= */ true);
        {
            let mut result = arb._query_external_agent("gpt-4o".to_string(), vec![HashMap::from([("role".to_string(), "user".to_string()), ("content".to_string(), "hi".to_string())])]).await;
        }
        assert!((result["content".to_string()].contains(&"ERROR".to_string()) || result["model".to_string()] == "gpt-4o".to_string()));
    }
    /// Test AutoGen swarm initialization stub.
    pub fn test_autogen_init(&self) -> Result<()> {
        // Test AutoGen swarm initialization stub.
        let mut arb = SwarmArbitrator(/* ports= */ vec![8001]);
        Ok(arb.init_autogen_swarm())
    }
}
